crates/compiler/builtins/roc/Str.roc

```diff
@@ -151,8 +151,91 @@ isEmpty : Str -> Bool
 concat : Str, Str -> Str
 
 ## Returns a string of the specified capacity without any content.
+##
+## This is a performance optimization tool that's like calling [Str.reserve] on an empty string.
+## It's useful when you plan to build up a string incrementally, for example by calling [Str.concat] on it:
+##
+## ```
+## greeting = "Hello"
+## subject = "World"
+##
+## # Evaluates to "Hello, World!\n"
+## helloWorld =
+##     Str.withCapacity 14
+##     |> Str.concat greeting
+##     |> Str.concat ", "
+##     |> Str.concat subject
+##     |> Str.concat "!\n"
+## ```
+##
+## Here, calling `"" |> Str.reserve 14` instead of `Str.withCapacity 14` would effectively do the same thing.
+##
+## In general, if you plan to use [Str.concat] on an empty string, it will be faster to start with
+## [Str.withCapacity] than with `""`. Even if you don't know the exact capacity of the string, giving [withCapacity]
+## a higher value than ends up being necessary can help prevent reallocation and copying—at
+## the cost of using more memory than is necessary.
+##
+## For more details on how the performance optimization works, see [Str.reserve].
 withCapacity : Nat -> Str
 
+## Increase a string's capacity by at least the given number of additional bytes.
+##
+## This can improve the performance of string concatenation operations like [Str.concat] by
+## allocating extra capacity up front, which can prevent the need for reallocations and copies.
+## Consider the following example which does not use [Str.reserve]:
+##
+## ```
+## greeting = "Hello"
+## subject = "World"
+##
+## # Evaluates to "Hello, World!\n"
+## helloWorld =
+##     greeting
+##     |> Str.concat ", "
+##     |> Str.concat subject
+##     |> Str.concat "!\n"
+## ```
+##
+## In this example:
+## 1. We start with `greeting = "Hello"`, which has both a length and capacity of 5 (bytes).
+## 2. `|> Str.concat ", "` will see that there isn't enough capacity to add 2 more bytes for the `", "`, so it will create a new heap allocation with enough bytes to hold both. (This probably will be more than 7 bytes, because when [Str] functions reallocate, they apply a multiplier to the exact capacity required. This makes it less likely that future realloctions will be needed. The multiplier amount is not specified, because it may change in future releases of Roc, but it will likely be around 1.5 to 2 times the exact capacity required.) Then it will copy the current bytes (`"Hello"`) into the new allocation, and finally concatenate the `", "` into the new allocation. The old allocation will then be deallocated because it's no longer referenced anywhere in the program.
```

Comment:
Is this a heap allocation? I thought it would be a small string.

Reply:
Oops, good call! Fixed in dca2d2afd81271e7ab0dd73a6bacc061df212a96